waspc/data/Generator/templates/server/src/routes/auth/signup.js

```diff
@@ -1,11 +1,23 @@
 {{={= =}=}}
+import Prisma from '@prisma/client'
 import prisma from '../../dbClient.js'
 import { handleRejection } from '../../utils.js'
+import EntityValidationError from '../../core/EntityValidationError.js'
+import HttpError from '../../core/HttpError.js'
 
 export default handleRejection(async (req, res) => {
   const userFields = req.body || {}
 
-  await prisma.{= userEntityLower =}.create({ data: userFields })
+  try {
+    await prisma.{= userEntityLower =}.create({ data: userFields })
+    res.send()
+  } catch(e) {
+    if (e instanceof Prisma.PrismaClientKnownRequestError) {
```

Comment:
this is cool, I haven't yet used Prisma errors, just now read more about it here: https://www.prisma.io/docs/reference/api-reference/error-reference

I haven't looked up the details yet but wanted to ask - why are we forwarding for `PrismaClientKnownRequestError` to the client but for others (e.g. `PrismaClientValidationError`) not (all other instances go to 500)?

Reply:
That is a good point. There was not a proper reason to only check for that class of error there; let me check the class hierarchy and see if we can just catch all possible Prisma errors on that line instead. Thanks! 👍🏻

Reply:
Doesn't seem so. Since `PrismaClientKnownRequestError` is the only one with a `code`, I will leave that as is, but add a new case handling the two other possible types: `PrismaClientUnknownRequestError` and `PrismaClientValidationError` and return their `message`.

`PrismaClientRustPanicError` and `PrismaClientInitializationError ` indicate more widespread issues we should probably handle elsewhere.

Reply:
Slight inconsistency here, you have:
```
  if ...
    throw ...
  else if ...
    throw ...
  throw ...
```

So one time you use `else` for control flow, and other time you rely on `throw` from above.

I would say we should stick with one of those, so it would be either

```
  if ...
    throw ...
  else if ...
    throw ...
  else
    throw ...
```

or

```
  if ...
    throw ...
  if ...
    throw ...
  throw ...
```

I maybe like the last one the best, just because it is very obvious we are certainly throwing something, although the first one has nicer indentation :D.

Reply:
Also, same question as @matijaSos , I find it unclear here as to why are we passing on that specific Prisma error! Also, I am guessing that just returning Prisma code will not be very informative to the client side? Unless they know what that means (which means we have that logic in client) -> but that feels like server is spilling its implementation details?

Reply:
Good call on unifying the flow, I can update that. The only reason I was thinking sending the code down may be helpful is if the client wanted to act on it, for example, the unique constraint error code, to show the user something informative in the form like: "Please try a different email". I didn't think the server should divulge that info down the wire for security, but then again if the client says something like that too it is a moot point. 😅 

So just wondering aloud, what can we provide from the server that can help the client, but not so much it allows for checking if an email is already registered? Or do we just give a generic save failed?

Reply:
It may clean things up a lot if we really on care to inform the user about validation errors, and then anything else that went wrong from Prisma is just a 422 with no message. I don't know that any other error case can happen from that flow, so the 500 base case may be unnecessary altogether. I'll chew on it.

Reply:
Yes it is a good point! I agree, maybe the most fool-proof way to start with is what you suggested above - take care of validation errors and forward them, and for all the rest go with 422. Maybe I would even leave 500 just in case Prisma doesn't act as expected, but you could e.g. add a note above `// This should never happen.`. Or maybe there is some middleware above which will catch that, I am not sure how it works right now?

Reply:
Btw what you mentioned above about a unique email, is a great catch. If we go with only validation and all the rest 422, what would happen in that case, 422 right? We should be able to detect that case, so if nothing else we can open a separate issue for this and handle it separately, so we don't make this PR too wide.

Reply:
I think those are all good concerns!
My thinking in general:
- If we are dealing with an error we understand (in server), meaning we know how to capture/recognize it:
  - If it is sensitive, mask it (don't provide too much info).
  - If it is not sensitive, pass it through.
- If we can't recognize an error, mask it.

So basically only very specific errors are sometihng we can return here.

As for the existing email -> if email already exist, and somebody is trying to do a sign-up, I don't see how we can not tell them that email already exists. I understand that it is not great in the sense that they can learn which emails are registered, but what other error message can we return? If we don't inform them that email is taken, they will have no idea how to fix this hm. So I don't see another way really. Maybe we pretend we created an account and say that we sent an email to them? But that is also a bit bad, it can also be very annoying. I think we should check how others are doing it -> we could check couple of popular web apps/pages and see how they handle this.

Reply:
Thanks both of you. Yeah, that "how will they know?" even though its not great from a security perspective was what I was torn with. So it sounds like a decent option is to identify just the unique constraint on `User.email` and report it back to the client, but mask all other Prisma errors (until we identify others to recognize and forward).

Reply:
If we are not sure about it from the security perspective, we can open an issue to think bout it again later, and mark it with "security" label (if we don't have one, let's create one).

Reply:
This is resolved here: https://github.com/wasp-lang/wasp/pull/376/commits/b2b086cdee30c3c8ea75786698cc0aff40b8f1f1 I sent back information on the columns of unique constraint violations as well, so the user knows the email is in use. I think that is a good approach for now and over time we can consider alternative approaches as they arise. Thanks!